### ISSUE TITLE
Errors in post-command hook

### ISSUE THREAD

**Issue by stsquad** [sjbach/lusty-emacs#18]
I've just done a fresh set-up of my emacs (24.3.1) and ELPA packages and lusty explorer seems to have broken. I'm not sure if this is due to a subtle change to the handling of flet.

Anyway the messages I'm seeing are:

```
Error in post-command-hook (lusty--post-command-function): (void-function add-buffer-maybe)
Quit
Type C-x 1 to delete the help window.
Error in post-command-hook (lusty--post-command-function): (void-function iterate-non-dedicated-window)
completing-read-default: Command attempted to use minibuffer while in minibuffer
Updating buffer list...
Formats have changed, recompiling...done
Updating buffer list...done
Commands: m, u, t, RET, g, k, S, D, Q; q to quit; h for help
Error in post-command-hook (lusty--post-command-function): (void-function iterate-non-dedicated-window)
Quit
Mark saved where search started
Error in post-command-hook (lusty--post-command-function): (void-function iterate-non-dedicated-window)
```

**Comment by stsquad** [comment]
More errors from the log:

    Error in post-command-hook (lusty--post-command-function): (invalid-function (iterate-non-dedicated-window (start-win direction) (let ((iterating-p t) (next start-win)) (while iterating-p (setq next (if (eq direction :forward) (next-window next :skip-mini) (previous-window next :skip-mini))) (when (or (not (window-dedicated-p next)) (eq next start-win)) (setq iterating-p nil))) next)))

**Comment by sjbach** [comment]
Thanks.  Are you running the HEAD lusty-explorer.el?  Do you always see the invalid-function error, or only sometimes?

**Comment by stsquad** [comment]
Yeah I'm running head but I only see that from time to time. I have stopped using lusty-buffer-explorer though because some interaction between lusty and Emacs 24.5 has stopped it seeing the buffer list. This even occurs with my most minimal config.

**Comment by sjbach** [comment]
Okay, thanks.  For what it's worth, I run GNU Emacs 25.0.50.1 and see no issues with lusty-buffer-explorer.

**Comment by stsquad** [comment]
@sjbach Am I reading the code right? Is (lusty-lowest-window) only ever meant to be called for Emacs 23 or bellow? It gets called in my case because:

    (window-live-p root-window)

Isn't true.

**Comment by stsquad** [comment]
I'm not totally sure what the window-live-p test is trying to do. The frame-root-window won't be live (as it doesn't contain a buffer itself) but it can be split to create a new window for lusty's results.

**Comment by sjbach** [comment]
`lusty-lowest-window` is annotated as being needed for Emacs 23 only, so perhaps at the time of writing the expectation was that `(window-live-p root-window)` would be possibly false in older versions but always true in newer versions.  But the [Emacs Lisp documentation](https://www.gnu.org/software/emacs/manual/html_node/elisp/Windows-and-Frames.html), presumably up-to-date, is more conservative:

> The root node of a window tree is called the root window. It can be either a live window (if the frame has just one window), or an internal window.

Summary: I'm not sure what the code is supposed to be doing.

**Comment by wisaac** [comment]
Hi! I'm using emacs 27 on OSX and received the following error when using lusty-buffer-explorer

`Error in post-command-hook (lusty--post-command-function): (void-function add-buffer-maybe)
`
Using the debugging tool you recommended [here](https://github.com/sjbach/lusty-emacs/issues/3), I got the following traceback:

```
Debugger entered--Lisp error: (void-function add-buffer-maybe)
  add-buffer-maybe(#<window 4 on  *Minibuf-1*>)
  walk-windows(add-buffer-maybe nil visible)
  lusty-buffer-list()
  lusty-buffer-explorer-matches("")
  lusty-refresh-matches-buffer()
  #f(compiled-function () #<bytecode 0x404d1fcd>)()
  ad-Advice-lusty--post-command-function(#f(compiled-function () #<bytecode 0x404d1fcd>))
  apply(ad-Advice-lusty--post-command-function #f(compiled-function () #<bytecode 0x404d1fcd>) nil)
  lusty--post-command-function()
  read-from-minibuffer(">> " nil (keymap (24 keymap (101 . lusty-select-current-name) (100 . lusty-launch-dired)) (down . lusty-highlight-next) (up . lusty-highlight-previous) (right . lusty-highlight-next-column) (left . lusty-highlight-previous-column) (2 . lusty-highlight-previous-column) (6 . lusty-highlight-next-column) (18 . lusty-highlight-previous) (19 . lusty-highlight-next) (16 . lusty-highlight-previous) (14 . lusty-highlight-next) (remap keymap (delete-backward-char . lusty-delete-backward)) (9 . lusty-select-match) (13 . lusty-open-this) keymap (menu-bar keymap (minibuf "Minibuf" keymap (previous menu-item "Previous History Item" previous-history-element :help "Put previous minibuffer history element in the minibuffer") (next menu-item "Next History Item" next-history-element :help "Put next minibuffer history element in the minibuffer") (isearch-backward menu-item "Isearch History Backward" isearch-backward :help "Incrementally search minibuffer history backward") (isearch-forward menu-item "Isearch History Forward" isearch-forward :help "Incrementally search minibuffer history forward") (return menu-item "Enter" exit-minibuffer :key-sequence "\015" :help "Terminate input and exit minibuffer") (quit menu-item "Quit" abort-recursive-edit :help "Abort input and exit minibuffer") "Minibuf")) (10 . exit-minibuffer) (13 . exit-minibuffer) (7 . abort-recursive-edit) (C-tab . file-cache-minibuffer-complete) (9 . self-insert-command) (XF86Back . previous-history-element) (up . previous-line-or-history-element) (prior . previous-history-element) (XF86Forward . next-history-element) (down . next-line-or-history-element) (next . next-history-element) (27 keymap (114 . previous-matching-history-element) (115 . next-matching-history-element) (112 . previous-history-element) (110 . next-history-element))) nil buffer-name-history nil nil)
  completing-read-default(">> " internal-complete-buffer nil nil nil buffer-name-history nil nil)
  read-buffer(">> ")
  apply(read-buffer ">> " nil)
  lusty--run(read-buffer)
  lusty-buffer-explorer()
  funcall-interactively(lusty-buffer-explorer)
  #<subr call-interactively>(lusty-buffer-explorer nil nil)
  apply(#<subr call-interactively> lusty-buffer-explorer (nil nil))
  call-interactively@ido-cr+-record-current-command(#<subr call-interactively> lusty-buffer-explorer nil nil)
  apply(call-interactively@ido-cr+-record-current-command #<subr call-interactively> (lusty-buffer-explorer nil nil))
  call-interactively(lusty-buffer-explorer nil nil)
  command-execute(lusty-buffer-explorer)
```
Tried playing around with the dot el files and nothing happened. Any help would be great!

**Comment by sjbach** [comment]
Hi William, thanks for the backtrace.  Does the issue still happen if you use the version of lusty-explorer.el at head?  I'm hoping this was fixed by 303618cafa01da3c8f99da4849d3ddbdc146a5d1.

**Comment by wisaac** [comment]
I installed the lusty-explorer.el at head directly from the repo, as well as melpa and still, got the same error. For some reason, when I add noflet package to the dot el file I no longer get the error but the buffer doesn't pull up any buffers.
